File: Java/src/main/java/com/gildedrose/handler/AgedBrieHandler.java
package com.gildedrose.handler;

import com.gildedrose.Item;
import com.gildedrose.utility.ItemUtilities;

public class AgedBrieHandler implements ItemHandler {
    @Override
    public boolean handles(String itemName) {
        return "Aged Brie".equals(itemName);
    }

    @Override
    public void updateItem(Item item) {
        if (ItemUtilities.canIncreaseQuality(item)) {
            ItemUtilities.increaseQuality(item);
        }

        ItemUtilities.decreaseSellIn(item);

        if (ItemUtilities.isExpired(item) && ItemUtilities.canIncreaseQuality(item)) {
            ItemUtilities.increaseQuality(item);
        }
    }
}

File: TypeScript/app/handler/AgedBrieHandler.ts
import { Item } from '../gilded-rose';
import { ItemHandler } from './ItemHandler';
import { ItemUtils } from '../utility/ItemUtils';

export class AgedBrieHandler implements ItemHandler {
  handles(item: Item): boolean {
    return item.name === 'Aged Brie';
  }

  updateItemQuality(item: Item): void {
    if (ItemUtils.canIncreaseQuality(item)) {
      ItemUtils.increaseQuality(item);
    }

    ItemUtils.decreaseSellIn(item);

    if (ItemUtils.isExpired(item) && ItemUtils.canIncreaseQuality(item)) {
      ItemUtils.increaseQuality(item);
    }
  }
}
